### PR TITLE
Convert elbo into a class for compatibility with Pyro

### PR DESCRIPTION
**Pull request by fehiepsi**
As discussed in #378, it is better to use a generator to create loss_fn because
+ Users don't care about elbo's args, kwargs.
+ This allows us to set hyperparameters of these loss functions.

### CODE REVIEW

**Comment by neerajprad** [comment]
How about just using an`ELBO` class like in Pyro, which should probably address all of the issues you raised in #378 and be much more extensible?

**Comment by fehiepsi** [comment]
Yup, we can make elbo a class to match Pyro. It will be a very simple class, that works just like a function (no stuff like differentiable_loss, loss_and_grad, loss,...). But again, matching Pyro interface is a plus for users. :)

File: docs/source/svi.rst [hunk 1]
```diff
@@ -9,4 +9,15 @@ Stochastic Variational Inference (SVI)
 
 ELBo
 ----
-.. autofunction:: numpyro.infer.svi.elbo
+
```
Comment (neerajprad) [inline docs/source/svi.rst]: Not for this PR, but we should rename this `ELBO` too to match the rest of the codebase.

File: examples/minipyro.py [hunk 2]
```diff
@@ -1,14 +1,13 @@
...
-import jax
 from jax import random
 import jax.numpy as np
 from jax.random import PRNGKey
 
 import numpyro
 from numpyro import optim
 import numpyro.distributions as dist
-from numpyro.infer import SVI, elbo
+from numpyro.infer import SVI, Trace_ELBO
```
Comment (neerajprad) [inline examples/minipyro.py]: We are stuck with the `Trace_ELBO` name in Pyro due to backwards compatibility reasons. Let us put this in compat if we need to do this for Pyro compatibility, and use the `ELBO` name in the main codebase.

File: numpyro/infer/elbo.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        :param args: arguments to the model / guide (these can possibly vary during
+            the course of fitting).
+        :param kwargs: keyword arguments to the model / guide (these can possibly vary
+            during the course of fitting).
+        :return: negative of the Evidence Lower Bound (ELBo) to be minimized.
+        """
+        raise NotImplementedError
+
+
+class Trace_ELBO(ELBO):
```
Comment (neerajprad) [inline numpyro/infer/elbo.py]: I would suggest merging this with `ELBO` and removing `Trace_ELBO`. Other classes are free to replace the loss method as they see fit.

File: examples/minipyro.py (hunk 2 above)
Comment (fehiepsi) [reply examples/minipyro.py]: done

File: docs/source/svi.rst (hunk 1 above)
Comment (fehiepsi) [reply docs/source/svi.rst]: done

File: numpyro/contrib/autoguide/__init__.py [hunk 4]
```diff
@@ -383,3 +388,42 @@ def _get_transform(self):
             arnn = numpyro.module('{}_arn__{}'.format(self.prefix, i), arn, (self.latent_size,))
             flows.append(InverseAutoregressiveTransform(arnn))
         return ComposeTransform(flows)
+
+
+class AutoContinuousELBO(ELBO):
```
Comment (neerajprad) [inline numpyro/contrib/autoguide/__init__.py]: So happy to see this refactored out!
